gtsfm/view_graph_estimator/view_graph_estimator_base.py

```diff
@@ -0,0 +1,123 @@
+"""Implements a base class for ViewGraph estimation.
+
+Estimating the ViewGraph can be done trivially by adding all the two-view estimates into a ViewGraph data structure.
+The purpose of this class, however, is to define an API for more sophisticated methods for estimating a ViewGraph 
+that include filtering or optimizing the two-view estimates.
+
+Authors: Akshay Krishnan, Ayush Baid
+"""
+import abc
+from typing import Dict, Optional, List, Tuple
+
+import dask
+import numpy as np
+from dask.delayed import Delayed
+from gtsam import Cal3Bundler, PinholeCameraCal3Bundler, Rot3, Unit3
+
+import gtsfm.utils.geometry_comparisons as comp_utils
+from gtsfm.evaluation.metrics import GtsfmMetric, GtsfmMetricsGroup
+from gtsfm.common.keypoints import Keypoints
+from gtsfm.common.view_graph import ViewGraph
+
+METRIC_GROUP = "view_graph"
+
+
+class ViewGraphEstimatorBase(metaclass=abc.ABCMeta):
+    """Base class for ViewGraph estimation.
+
+    A ViewGraphEstimator aggregates two-view estimates into a ViewGraph.
+    It could also improve the two-view estimates using filtering or optimization techniques.
+    """
+
+    @abc.abstractmethod
+    def run(
+        self,
+        i2Ri1: Dict[Tuple[int, int], Rot3],
+        i2Ui1: Dict[Tuple[int, int], Unit3],
+        calibrations: List[Cal3Bundler],
+        corr_idxs_i1i2: Dict[Tuple[int, int], np.ndarray],
+        keypoints: List[Keypoints],
+    ) -> ViewGraph:
+        """Run the ViewGraph estimation.
+
+        Args:
+            i2Ri1: Dict from (i1, i2) to relative rotation of i1 with respect to i2.
+            i2Ui1: Dict from (i1, i2) to relative translation direction of i1 with respect to i2.
+            calibrations: list of calibrations for each image.
+            corr_idxs_i1i2: Dict from (i1, i2) to indices of verified correspondences from i1 to i2.
+            keypoints: keypoints for each images.
+
+        Returns:
+            ViewGraph object.
+        """
+
+    def compute_metrics(
```

Comment:
I like the idea of having a standard set of metrics for each estimator, but it seems as though CycleConsistentRotationViewGraphEstimator just overwrites this. I think this can be implemented in such a way that it could capture all relevant metrics for any derived class.

Reply:
Yeah we can perhaps move those metrics here, and provide a method like _compute_additional_metrics that users can override to add more metrics if they like.

Reply:
If the derived implementation wants to override the method, they can call the super method. We can document it in the docstring. I am against having any *additional* function.